orderer/common/channelparticipation/restapi.go

```diff
@@ -127,17 +132,67 @@ func (h *HTTPHandler) serveListOne(resp http.ResponseWriter, req *http.Request)
 	h.sendResponseOK(resp, infoFull)
 }
 
-// Join a channel
+// Join a channel. Expect application/json content.
 func (h *HTTPHandler) serveJoin(resp http.ResponseWriter, req *http.Request) {
 	_, err := negotiateContentType(req) // Only application/json for now
 	if err != nil {
 		h.sendResponseJsonError(resp, http.StatusNotAcceptable, err)
 		return
 	}
 
-	//TODO
-	err = errors.Errorf("not implemented yet: %s %s", req.Method, req.URL.String())
-	h.sendResponseJsonError(resp, http.StatusNotImplemented, err)
+	channelID, ok := mux.Vars(req)[channelIDKey]
+	if !ok {
+		h.sendResponseJsonError(resp, http.StatusInternalServerError, errors.Wrap(err, "missing channel ID"))
+		return
+	}
+
+	if err = configtx.ValidateChannelID(channelID); err != nil {
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "invalid channel ID"))
+		return
+	}
+
+	bodyBytes, err := ioutil.ReadAll(req.Body)
+	if err != nil {
+		h.logger.Debugf("Failed to read request body: %s", err)
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "cannot read request body"))
+		return
+	}
+
+	body := types.JoinBody{}
+	err = json.Unmarshal(bodyBytes, &body)
+	if err != nil {
+		h.logger.Debugf("Failed to json.Unmarshal request body: %s", err)
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "cannot json.Unmarshal request body"))
+		return
+	}
+
+	block := &cb.Block{}
+	err = proto.Unmarshal(body.ConfigBlock, block)
+	if err != nil {
+		h.logger.Debugf("Failed to unmarshal ConfigBlock field: %s", err)
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "cannot unmarshal ConfigBlock field"))
+		return
+	}
+
+	info, err := h.registrar.JoinChannel(channelID, block)
+	if err == nil {
+		info.URL = path.Join(URLBaseV1Channels, info.Name)
+		h.logger.Debugf("Successfully joined channel: %s", info)
+		h.sendResponseCreated(resp, info.URL, info)
+		return
+	}
+
+	h.logger.Debugf("Failed to JoinChannel: %s", err)
+	switch err {
+	case types.ErrSystemChannelExists:
+		h.sendResponseNotAllowed(resp, errors.Wrap(err, "cannot join"), http.MethodGet)
+	case types.ErrChannelAlreadyExists:
+		h.sendResponseNotAllowed(resp, errors.Wrap(err, "cannot join"), http.MethodGet, http.MethodDelete)
+	case types.ErrAppChannelsAlreadyExists:
+		h.sendResponseJsonError(resp, http.StatusForbidden, errors.Wrap(err, "cannot join"))
+	default:
+		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "cannot join"))
+	}
```

Comment:
I think that since you treat `ErrSystemChannelExists` And `ErrChannelAlreadyExists` you might want to consolidate it into

```
   case types.ErrSystemChannelExists, types.ErrChannelAlreadyExists: {
   } 
```

Also I think that in general it might be good to extract error treatment into separate function.

Reply:
They are somewhat different, as the "Allow" header in the response is different:
* In the `ErrSystemChannelExists` case you can only GET the channel (whether it exists or not is not checked here).
* In the `ErrChannelAlreadyExists` case, the system channel does not exist, but the channel itself does. This means that what you can do on the channel is GET & DELETE.

```go
switch err {
	case types.ErrSystemChannelExists:
		h.sendResponseNotAllowed(resp, errors.Wrap(err, "cannot join"), 
                http.MethodGet)
	case types.ErrChannelAlreadyExists:
		h.sendResponseNotAllowed(resp, errors.Wrap(err, "cannot join"), 
                http.MethodGet, http.MethodDelete)
	case types.ErrAppChannelsAlreadyExists:
		h.sendResponseJsonError(resp, http.StatusForbidden, errors.Wrap(err, "cannot join"))
	default:
		h.sendResponseJsonError(resp, http.StatusBadRequest, errors.Wrap(err, "cannot join"))
	}
```

I will restructure the error handling to a separate function.